lektor/admin/static/js/main.tsx

```diff
@@ -12,63 +14,70 @@ import DeletePage from "./views/delete/DeletePage";
 import PreviewPage from "./views/PreviewPage";
 import AddChildPage from "./views/add-child-page/AddChildPage";
 import AddAttachmentPage from "./views/AddAttachmentPage";
-import { getRecordDetails } from "./components/RecordComponent";
 import { adminPath } from "./components/use-go-to-admin-page";
+import { trimSlashes } from "./utils";
 
-setCurrentLanguage($LEKTOR_CONFIG.lang);
+type ViewComponentProps = Pick<RecordProps, "record">;
 
-function getMainComponent(page: string) {
-  if (page === "edit") {
-    return EditPage;
-  } else if (page === "delete") {
-    return DeletePage;
-  } else if (page === "preview") {
-    return PreviewPage;
-  } else if (page === "add-child") {
-    return AddChildPage;
-  } else if (page === "upload") {
-    return AddAttachmentPage;
-  }
-  return null;
+interface MatchParams {
+  path?: string | undefined;
 }
 
-function Main() {
-  const match = useRouteMatch<{ path: string; page: string }>(
-    `${$LEKTOR_CONFIG.admin_root}/:path/:page`
-  );
-  // useRouteMatch returns a new object on each render, so we need to get the
-  // primitive string values here to memoize.
-  const urlPath = match?.params.path;
-  const page = match?.params.page;
+interface LayoutProps extends RouteComponentProps<MatchParams> {
+  page: string;
+  component: ComponentType<ViewComponentProps>;
+}
 
-  const record = useMemo(() => {
-    if (!urlPath) {
-      return null;
-    }
-    return getRecordDetails(urlPath);
-  }, [urlPath]);
+function Layout({ match, page, component: Component }: LayoutProps) {
+  const record = {
+    path: `/${trimSlashes(decodeURIComponent(match.params.path ?? ""))}`,
+    alt: useRecordAlt(),
+  };
 
-  if (!page) {
-    return <Redirect to={adminPath("edit", "/", "_primary")} />;
-  }
-  const Component = getMainComponent(page);
-  if (!Component || record === null) {
-    return <Redirect to={adminPath("edit", "/", "_primary")} />;
-  }
   return (
     <App page={page} record={record}>
       <Component record={record} />
     </App>
   );
 }
 
-const dash = document.getElementById("dash");
+function Main() {
+  const admin_root = $LEKTOR_CONFIG.admin_root;
+  const views = [
+    { page: "edit", component: EditPage },
+    { page: "delete", component: DeletePage },
+    { page: "preview", component: PreviewPage },
+    { page: "add-child", component: AddChildPage },
+    { page: "upload", component: AddAttachmentPage },
+  ];
 
-if (dash) {
-  ReactDOM.render(
+  const routes = views.map((view) => (
+    <Route
+      key={view.page}
+      path={`${admin_root}/${view.page}/:path*`}
+      exact
+      render={(props) => <Layout {...props} {...view} />}
+    />
+  ));
+
+  const fallbackRoute = (
+    <Route path="*">
+      <Redirect to={adminPath("edit", "/", "_primary")} />
+    </Route>
+  );
+
+  return (
     <BrowserRouter>
-      <Main />
-    </BrowserRouter>,
-    dash
+      <Switch>
```

Comment:
I've just refactored this again, adding stronger type checking, etc.  Take a look.  I really think it's cleaner now.

There has to be a switch on page name somewhere.  It might as well be in a `<Switch>`.  This eliminates to need to handle invalid URLs in multiple places.

(The stuff in RecordComponent.tsx could certainly be refactored.)

My new `Page` component (in main.tsx) is responsible for grabbing the record path and alt from the route match and query params.  If you want to add a context provider, that's probably where to hook it in.

Reply:
I like it